### PR TITLE
Implement new SignedExtension for checking PoV sizes for every block

### PR DESCRIPTION
**Pull request by KiChjang**
This PR creates a new `SignedExtension` called `CheckPov` which is very similar to `CheckWeight`, but instead checks for the PoV size limits for each block instead.

Every new dispatch function will now have an additional attribute called `#[pallet::pov_size(expr)]` similar to `#[pallet::weight(expr)]`, which is used to provide an expression that evaluates to a `u32` representing the PoV size used for the dispatch.

A new Config item called `PovParams` is created for frame system, and most of the time its value should be set according to the `AbridgedHostConfiguration` struct that the relay chain provides.

Runtime implementors who want to check for PoV size limits will have to remember to add `CheckPov` as a `SignedExtension` when they declare their runtime.

polkadot companion: paritytech/polkadot#3839

### CODE REVIEW

File: frame/system/src/extensions/check_pov.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	traits::Get,
+	weights::{DispatchInfo, PostDispatchInfo},
+};
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, SignedExtension},
+	transaction_validity::{InvalidTransaction, TransactionValidityError},
+};
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, Default)]
+pub struct CheckPov<T: Config + Send + Sync>(sp_std::marker::PhantomData<T>);
```
Comment (kianenigma) [inline frame/system/src/extensions/check_pov.rs:30]: pub item should be documented.

**Review by kianenigma — COMMENTED** [review]
The code looks overall good

My main note is about: do we assume that PoV is a thing in all frame based chain, or should we try and hide it as much as possible? Imagine you are a solo-chain, and you genuinely don't care about PoV size. I presume it should be easy to opt out of everthing

File: frame/support/src/dispatch.rs [hunk 2]
```diff
@@ -2097,6 +2097,7 @@ macro_rules! decl_module {
 					$(
 						$call_type::$fn_name( $( ref $param_name ),* ) => {
 							let __pallet_base_weight = $weight;
+							let __pallet_pov_size = 0;
```
Comment (kianenigma) [inline frame/support/src/dispatch.rs:2100]: yes, let's not support stuff here finally :D

**Review by kianenigma — APPROVED** [review]

**Comment by thiolliere** [comment]
> > My main note is about: do we assume that PoV is a thing in all frame based chain, or should we try and hide it as much as possible? Imagine you are a solo-chain, and you genuinely don't care about PoV size. I presume it should be easy to opt out of everthing
> 
> We should make it possible to hide PoV from your runtime development if it is not relevant to you, i.e. a solo chain.
> 
> Yes I agree, should be easy to opt out of this stuff.

In this PR people can just use random values and not use the extension which check PoV size. Which sounds ok.
But it would be cleaner to create another pallet which just check the PoV size maybe, I don't see the need for it to be in frame_system. frame_pov would be fine. Anyhow the PR is fine to me

File: frame/support/src/dispatch.rs (hunk 2 above)
Comment (KiChjang) [reply frame/support/src/dispatch.rs:2100]: Lol yes, I got too lazy to support any more new syntax there, so I stubbed this out with just 0.

**Comment by bkchr** [comment]
Can someone shed some light on this on how this should work? As the PoV size configured by the relay chain is for a compressed PoV, does that mean we will also zstd compress the tx + proof while benchmarking to calculate an upper limit that will be put into the attribute above each extrinsic?

**Comment by KiChjang** [comment]
> Can someone shed some light on this on how this should work? As the PoV size configured by the relay chain is for a compressed PoV, does that mean we will also zstd compress the tx + proof while benchmarking to calculate an upper limit that will be put into the attribute above each extrinsic?

That is pretty much the flow, though I'm not 100% sure about the details such as whether it's compressed, but essentially all PoV limits come from the relay chain, while the PoV size usage of each extrinsic is determined via benchmarking, similar to the infrastructure behind determining weights.

**Comment by athei** [comment]
This might be of scope here but shouldn't transactions also be charged fees based upon PoV as it is the case for weight?

File: frame/support/src/weights.rs [hunk 3]
```diff
@@ -320,6 +320,8 @@ pub struct DispatchInfo {
 	pub class: DispatchClass,
 	/// Does this transaction pay fees.
 	pub pays_fee: Pays,
+	/// PoV size used by this transaction.
+	pub pov_size: u32,
```
Comment (thiolliere) [inline frame/support/src/weights.rs:324]: related to Basti question https://github.com/paritytech/substrate/pull/9763#issuecomment-918458515
Here is should be written what is the PoV size, either it correspond to the compressed size of the PoV for the given dispatchable or the non-compressed one.

File: bin/node-template/runtime/src/lib.rs [hunk 4]
```diff
@@ -149,6 +150,8 @@ impl frame_system::Config for Runtime {
 	type BlockWeights = BlockWeights;
 	/// The maximum length of a block (in bytes).
 	type BlockLength = BlockLength;
+	/// The PoV parameters of a block. Values should be fetched from the relay chain in production.
+	type PovParams = PovParams;
```
Comment (thiolliere) [inline bin/node-template/runtime/src/lib.rs:154]: for solo chain maybe we can say that the value should be "unlimited".

File: frame/system/src/limits.rs [hunk 5]
```diff
@@ -429,6 +429,21 @@ impl BlockWeightsBuilder {
 	}
 }
 
+/// Proof-of-Validity (PoV) configuration parameters.
+#[derive(RuntimeDebug, Clone, codec::Encode, codec::Decode)]
+pub struct PovParams {
+	/// Maximum PoV size in bytes available in each block.
```
Comment (thiolliere) [inline frame/system/src/limits.rs:435]: same we should probably say if it the compressed or uncompressd size I guess

File: frame/support/procedural/src/pallet/parse/call.rs [hunk 6]
```diff
@@ -55,14 +56,17 @@ pub struct CallVariantDef {
...
+	pub pov_size: Option<syn::Expr>,
 	/// Docs, used for metadata.
 	pub docs: Vec<syn::Lit>,
 }
 
 /// Attributes for functions in call impl block.
-/// Parse for `#[pallet::weight(expr)]`
-pub struct FunctionAttr {
-	weight: syn::Expr,
+/// Parse for `#[pallet::weight(expr)]` or `#[pallet::pov_size(expr)]`
```
Comment (thiolliere) [inline frame/support/procedural/src/pallet/parse/call.rs:66]: currently `pallet::weight` give all the dispatch info except pov_size. I know its name is weird, but maybe we should merge everything into one attribute ?
but it is probably better as you do now.

**Comment by thiolliere** [comment]
We should also be able to return some PoV size in on_initialize and in on_runtime_upgrade isn't it ?

**Comment by apopiak** [comment]
nitpick: PoV is a Polkadot term. Is it the best name for the generic Substrate concept?
Naming Ideas: StorageDelta, ChangeSet

**Comment by thiolliere** [comment]
do we also want to have refund for PoV size ?

**Comment by thiolliere** [comment]
> nitpick: PoV is a Polkadot term. Is it the best name for the generic Substrate concept?
> Naming Ideas: StorageDelta, ChangeSet

true, either we are abstract on this or we consider it PoV and people can opt out not use it (by setting no limit to PoV)

**Comment by athei** [comment]
> do we also want to have refund for PoV size ?

Yes. I don't see how contracts would work without it. I think we need to treat it like gas there. Have a `pov_limit` argument to the call and meter the access to storage.

File: frame/system/src/extensions/check_pov.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		crate::PovSize::<T>::put(next_size);
+		Ok(())
+	}
+
+	fn check_pov_size(
+		info: &DispatchInfoOf<T::Call>,
+	) -> sp_std::result::Result<u32, TransactionValidityError> {
+		let size_limit = T::PovParams::get().max_size;
+		let current_size = Pallet::<T>::pov_size();
+		let added_size = info.pov_size;
```
Comment (bkchr) [inline frame/system/src/extensions/check_pov.rs:61]: What is with the size of the extrinsic itself? It also contributes to the size of the PoV.

**Review by bkchr — COMMENTED** [review]
https://github.com/paritytech/substrate/blob/master/client/basic-authorship/src/basic_authorship.rs#L378-L394

Here we do the checking of the PoV size on the node side. But be aware that this still does not includes any compression.

**Pull request closed without merging**